coderd/debug.go

```diff
@@ -8,20 +8,85 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
 	"golang.org/x/exp/slices"
 	"golang.org/x/xerrors"
 
-	"github.com/google/uuid"
-
 	"github.com/coder/coder/v2/coderd/audit"
 	"github.com/coder/coder/v2/coderd/database"
 	"github.com/coder/coder/v2/coderd/healthcheck"
 	"github.com/coder/coder/v2/coderd/httpapi"
-	"github.com/coder/coder/v2/coderd/httpmw"
 	"github.com/coder/coder/v2/coderd/rbac"
 	"github.com/coder/coder/v2/codersdk"
 )
 
+type debugHealthMetrics struct {
+	accessURLHealthyGauge        *prometheus.GaugeVec
+	accessURLReachableGauge      *prometheus.GaugeVec
+	accessURLStatusCodeGauge     *prometheus.GaugeVec
+	accessURLResponseLengthGauge *prometheus.GaugeVec
+	databaseHealthGauge          *prometheus.GaugeVec
+	databaseReachableGauge       *prometheus.GaugeVec
+	databaseLatencyGauge         *prometheus.GaugeVec
+	databaseThresholdGauge       *prometheus.GaugeVec
+	derpHealthyGauge             *prometheus.GaugeVec
+	derpNodeHealthyGauge         *prometheus.GaugeVec
+	derpNodeRoundTripPingGauge   *prometheus.GaugeVec
+	derpNodeUsesWebsocketGauge   *prometheus.GaugeVec
+	derpNodeStunEnabledGauge     *prometheus.GaugeVec
+	websocketHealthyGauge        *prometheus.GaugeVec
+	websocketResponseLengthGauge *prometheus.GaugeVec
+	websocketStatusCodeGauge     *prometheus.GaugeVec
+}
+
+func (api *API) DebugHealthcheckLoop(ctx context.Context, refresh time.Duration) {
+	ticker := time.NewTicker(refresh)
+	defer ticker.Stop()
+
+	// Run the first healthcheck on startup.
+	r := api.HealthcheckFunc(ctx)
+	api.healthCheckCache.Store(r)
+	api.reportHealthcheckPrometheusMetrics(r)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			r := api.HealthcheckFunc(ctx)
+			api.healthCheckCache.Store(r)
+			api.reportHealthcheckPrometheusMetrics(r)
+		}
+	}
+}
+
+func (api *API) reportHealthcheckPrometheusMetrics(report *healthcheck.Report) {
+	// Access URL
+	api.debugHealthMetrics.accessURLHealthyGauge.WithLabelValues().Set(boolToFloat64(report.AccessURL.Healthy))
+	api.debugHealthMetrics.accessURLReachableGauge.WithLabelValues().Set(boolToFloat64(report.AccessURL.Reachable))
+	api.debugHealthMetrics.accessURLStatusCodeGauge.WithLabelValues().Set(float64(report.AccessURL.StatusCode))
+	api.debugHealthMetrics.accessURLResponseLengthGauge.WithLabelValues().Set(float64(len(report.AccessURL.HealthzResponse)))
+	// Database
+	api.debugHealthMetrics.databaseHealthGauge.WithLabelValues().Set(boolToFloat64(report.Database.Healthy))
```

Comment:
Hey, FYI - we have deprecated `Healthy bool` property in favor of `Severity`. It would be convenient to depend on the Severity property which is more descriptive (ok, warning, error) 👍

Reply:
Sounds good, thanks for the heads up. I can only report ints - would it make sense to map this to severity 0,1,2?

Reply:
Yes, it seems to be a reasonable choice.